### ISSUE TITLE
False positive for object_usage_linter

### ISSUE THREAD

**Issue by jabenninghoff** [r-lib/lintr#1088]
Until recently, the test_that code below returned no lints. Recently, this code began returning errors for object_usage_linter:

*(This is an abbreviated example from [test-release.R](https://github.com/jabenninghoff/rdev/blob/main/tests/testthat/test-release.R))*

```r
test_that("abbriviated example", {
  gh_pulls <- list(list(title = "testpkg 1.2.0"))
  gh_pull_number <- list(
    locked = FALSE, draft = FALSE, mergeable = TRUE, rebaseable = TRUE,
    html_url = "https://github.com/example/test"
  )
  gh_merge <- list(merged = TRUE)
  gh <- function(command, ...) {
    if (command == "GET /repos/{owner}/{repo}/pulls") {
      return(gh_pulls)
    }
    if (command == "GET /repos/{owner}/{repo}/pulls/{pull_number}") {
      return(gh_pull_number)
    }
    if (command == "PUT /repos/{owner}/{repo}/pulls/{pull_number}/merge") {
      return(gh_merge)
    }
  }
})
```

Expected result: no lints
Actual result:
```
no visible binding for global variable ‘gh_pulls’
no visible binding for global variable ‘gh_pull_number’
no visible binding for global variable ‘gh_merge’
```

The issue appears to be introduced by #1082.

**Comment by MichaelChirico** [comment]
Thanks for early testing! I added this bug to the upcoming release to ensure it gets fixed.

**Comment by AshesITR** [comment]
The root cause is `get_function_assignments()` no longer only returns top-level function assignments due to the `//*` XPath.
This means we run `checkUsage()` on more functions than before.
In order to be useful for non-top-level functions, the list of known symbols would also need to be correctly populated for nested functions, i.e., we'd have to compute a list of known variables for each linted function individually.

As a quick fix for 3.0.0 I suggest reverting back to only linting top-level functions and creating an issue to also lint nested functions.
